File: 프로그래머스/unrated/172928. 공원 산책/공원 산책.py
def solution(park, routes):
    dx = {'N':-1 ,'S':1 ,'W':0 ,'E':0 }
    dy = {'N':0 ,'S':0 ,'W':-1 ,'E':1 }
    
    # 시작좌표 찾기(x, y)
    x = -1
    for p in park:
        x += 1
        if 'S' in p:
            y = p.index('S')
            break
            
    # 방향 이동(한 칸씩 이동 -> 조건 만족 시 명령 수행)
    for r in routes:
        op, n = r.split()
        for i in range(1, int(n)+1):
            nx = x + i*dx[op]
            ny = y + i*dy[op]
            # 이동할 수 있다면 현재 방향으로 계속 진행
            if 0 <= nx < len(park) and 0 <= ny < len(park[0]) and park[nx][ny] != 'X':
                continue
            # 이동할 수 없다면 현재 방향 무시
            else:
                break
                
        # 이동 방향 업데이트
        else:
            x, y = nx, ny
        
    return [x, y]

File: 프로그래머스/1/172928. 공원 산책/공원 산책.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

vector<int> solution(vector<string> park, vector<string> routes) {
    int H = park.size();
    int W = park[0].size();
    
    map<char, pair<int, int>> mp;
    mp['N'] = {-1, 0};
    mp['S'] = {1, 0};
    mp['W'] = {0, -1};
    mp['E'] = {0, 1};
    
    pair<int, int> loc;
    for (int i=0; i<H; i++) {
        for (int j=0; j<W; j++) {
            if (park[i][j] == 'S') loc = {i, j};
        }
    }
    
    for (const auto &route: routes) {
        char dir = route[0];
        int step = route[2] - '0';
        
        int ny = loc.first;
        int nx = loc.second;
        while(step--) {
            ny += mp[dir].first;
            nx += mp[dir].second;
            if (ny < 0 || nx < 0 || ny >= H || nx >= W) break;
            if (park[ny][nx] == 'X') break;
        }
        if (step < 0) loc = { ny, nx };
    }
    return { loc.first, loc.second };
    
}